server/aesdsocket.c

```diff
@@ -429,25 +470,56 @@ int main(int argc, char **argv)
     	if(thread_args==NULL)
     	{
     		syslog(LOG_ERR,"memory allocation for thread arguments failed");
-    		close(data_fd);
+    		close(fd);
     		continue;
     	}
     	thread_args->fd=fd;
-    	thread_args->data_fd=data_fd;
     	thread_args->addr=client_addr;
+#if USE_AESD_CHAR_DEVICE
+        thread_args->data_fd = open(FILE_NAME, O_RDWR);
+#else
+        thread_args->data_fd = data_fd;  // Use the same `data_fd` file descriptor if not using aesdchar
+#endif
+
+        if (thread_args->data_fd == -1) {
+            syslog(LOG_ERR, "Failed to open %s", FILE_NAME);
+            close(fd);
+            free(thread_args);
+            continue;
+        }
+    	
     	if(pthread_create(&thread_id,NULL,thread_function,(void *)thread_args) !=0)
     	{
     		syslog(LOG_ERR,"failed to create timer thread");
-    		free(thread_args);
+    		
     		close(data_fd);
+    		free(thread_args);
     		continue;
     	}
     	add_new_thread_node(thread_id);
+    	wait_for_all_threads_to_join();
```

Comment:
@shruthi-thallapally why move this into the while loop?  It will mean you now will only allow a single thread to run.